tests/run/i24357.scala

```diff
@@ -0,0 +1,10 @@
+class E1 extends Exception
+class E2 extends Exception
+
+type E1or2 = E1 | E2
+
+@main def Test =
+    try throw new Exception {}
+    catch
+      case e: E1or2 => assert(e.isInstanceOf[E1or2])
+      case _ => ()
```

Comment:
This makes the test more robust.
```suggestion
      case e: E1or2 => assert(false)
      case _ => assert(e.isInstanceOf[E1or2])
```